### PR TITLE
Arc support

### PR DESCRIPTION
**Pull request by achmed20**
module: ARC_SUPPORT, Adds G2/G3 (IJ version) support to klipper

G2/G3 commands are converted into G1 moves and added to the queue. Steps for the arc's are done in mm set in config. The  function used is pulled and converted from Marlin. 

My first ever pull request, so please be kind

Signed-off-by: Aleksej Vasiljkovic <achmed21@gmail.com>

### CODE REVIEW

**Comment by achmed20** [comment]
prrof of function ;)

**Comment by KevinOConnor** [comment]
Thanks.  FYI, I wont be able to take a detailed look at this for a week or so.

-Kevin

File: config/example-extras.cfg [hunk 1]
```diff
@@ -1792,3 +1792,14 @@
 #   A list of G-Code commands to execute when the delay duration has
 #   elapsed.  G-Code templates are supported.  This parameter must be
 #   provided.
+#
+# enables arc commands. G2/G3 commands will be converted into G1 codes
+# Only IJ version is supported
+# example: "G2 X125 Y32 Z10 E5 I10.5 J10.5"
+#[arc_support]
+#mm_per_step: 1.0
```
[the remark this answers is not in the document]
Comment (achmed20) [reply config/example-extras.cfg]: I just renamed it to "resolution" and changed the description to make it more clear

File: klippy/extras/arc_support.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
+# adds g2/g3 support. converts both commands into g1 code
+#
+# uses the plan_arc function from marlin which does steps in mm rather then
+# in degrees.
+# Coordinates created by this are converted into G1 commands. It is a bit
+# of a lazy aproach but i didnt feel like copying the G1 code functions.
+#
+# note: only IJ version available
+#
+# Copyright (C) 2019  Aleksej Vasiljkovic <achmed21@gmail.com>
```
[the remark this answers is not in the document]
Comment (achmed20) [reply klippy/extras/arc_support.py]: not realy sure how i should handle this. i changed it and added the source to marlin. is this enough?

[the remark this answers is not in the document]
Comment (KevinOConnor) [reply klippy/extras/arc_support.py]: Preference would be:
```
# Support for G2/G3 g-code commands
#
# Copyright (C) 2019  Aleksej Vasiljkovic <achmed21@gmail.com>
# Copyright (C) 2011 Camiel Gubbels / Erik van der Zalm
#
# This file may be distributed under the terms of the GNU GPLv3 license.

```

But what you have is also okay.

**Comment by achmed20** [comment]
all done btw. Is there anything else i need to do here to get this merged?

**Comment by KevinOConnor** [comment]
Thanks.  I made a minor change (I renamed the file to gcode_arcs.py) and committed this (commit 7d8c7036).

I think it would be worthwhile to follow up with the float() changes I previously recommended.  Basically any call to float() either needs to be wrapped in a try/except block, or one of the getfloat() helper functions should be used.  Otherwise, a typo in config or commands will result in an internal error (instead of an appropriate user facing error message).

-Kevin

**Comment by Adirael** [comment]
Hello.

I'm trying this out and it doesn't seem to work for me. I'm running the latest master and the Z portion of the G3 instruction is being interpreted as a relative movement. 

The following gcode portion will home, set movement to absolute positioning, set the units to mm and go to Z11.5. Then the toolhead is moved to X103 Y53 Z8.5.

The next command (G2) will drop the toolhead another 8.5mm instead of leaving it where it is. 

I'll try to debug it but I don't know a lot of Python, maybe someone can take a look at my gcode before I dig into this to confirm the instructions are correct?

Thanks! 

> G28 
> G90
> G21 
> 
> G0 Z11.500000
> G0 X103.355718 Y53.542738
> 
> G1 Z8.500000 F2500.0
> G1 X98.903198 Y53.542738 Z8.500000 F2500.000000
> G2 X98.610309 Y53.664056 Z8.500000 I0.000000 J0.414207
> G2 X98.488991 Y53.956945 Z8.500000 I0.292889 J0.292889

**Comment by achmed20** [comment]
> Hello.
> 
> I'm trying this out and it doesn't seem to work for me. I'm running the latest master and the Z portion of the G3 instruction is being interpreted as a relative movement.
> ...
checked that and it was indeed wrong (leftovers from some old code).  that only happend when you actualy included the Z coordinate in the g2/g3 code which none of my generators (inkscape) actualy ever did :)

I removed using calculated  Z height entirely and instead use a the one defined in G2/G3 (if present). it doesnt seem like Arcs were ever supposed to be 3D anyway.

@KevinOConnor 
I merged your master back into my fork and cleaned it up, so you should be able to pull these changes.

**Comment by KevinOConnor** [comment]
Thanks.  I'm not sure I understand the `resolution` change - why can't a user specify a finer resolution if they wish?  Also, the caller shouldn't catch and rethrow the error - the getfloat() method automatically produces a consistent error message.

-Kevin

**Comment by achmed20** [comment]
concerning the resolution
1. 0.1mm is allready very small and you wouldnt be able to notice this on a print/plot anyway
2. because a full circle with a radius of 100mm and 0.1mm arc segments is allready 6282 gcode steps.

for the try/catch
mostly because you said so and also because i never did get any kind of error if i mess up the values in the config (like writing something like "1mm"). it jsut failes to load without staing why and where.

**Comment by KevinOConnor** [comment]
Every `float()` call should either be wrapped in a try/except block or converted to a getfloat() call.  The `getfloat()` calls should not be wrapped.

The host machine is very fast and some people perform very detailed prints.  I recommend changing the restriction to `above=0.0` and not specify a minval.

-Kevin

**Comment by achmed20** [comment]
done

**Comment by KevinOConnor** [comment]
Thanks.

-Kevin

**Comment by leonmelson** [comment]
Hi I am also trying the arcs and it works fine if all XY and IJ values are provided.

I have found that some cad programs do not provide one of the XY or IJ values if they are equal to the previous value.

Example
G1 X18.557 Y27.148
G3 X18.999 Y26.969 I0.348 J0.224
G3 X19.309 Y27.332 I-0.091 J0.392
G3 X19.254 Y27.533 I-0.315 J0.023
G2 Y67.467 I29.013 J19.967

Not sure if this is something that you would like to add or just leave for people to fix in there post processors.

Thanks you for the awesome work on this.

**Pull request merged**